[Step2] 문자열 계산기

loadkrnis opened the pull request:
안녕하세요 :)
즐거운 어린이날, 어버이날 보내셨나요~?
저는 정신없이 돈만 쓴 것 같네요 하하 😅

이번 step2는 테스트를 먼저 작성하고 구현을 해보았는데요.
이렇게 완전 TDD로 개발을 해본건 처음이라 궁금한 점이 생겨서 리뷰어님의 의견이 궁금합니다 :)

먼저 테스트를 작성 후 기능을 구현하게 되면서 기존에 짰던 테스트를 일부 수정하게 되는 일이 몇번 있었는데요.
(ex. public 메소드에서 private 메소드로 변경, instance 메소드에서 static 메소드로 변경 등)
구현중에 위와 같은 테스트 변경은 어쩔 수 없는 건지 궁금합니다. 

처음에 테스트를 작성할 때 
앞으로 이 메소드를 static으로 사용할지, 
instance 메소드로 사용할지 
먼저 결정을 해야하는걸까요?
테스트를 짜는데 자꾸 구현부를 생각하는 것 같아서 TDD가 맞나? 라는 생각이 들기도 했고, 
막상 어느정도 그림을 그려가며 테스트를 짜니까 구현에 도움이 된 것 같다는 생각도 들었습니다.

리뷰어님의 생각은 어떠신지 궁금합니다 :)
금주도 화이팅입니다! 💪

loadkrnis committed:
[step2] 요구 사항 정리

loadkrnis committed:
[step2] 사칙연산 테스트 추가

loadkrnis committed:
[step2] 사칙연산 기능 추가

loadkrnis committed:
[step2] 숫자, 연산자 분리기 테스트 추가

loadkrnis committed:
[step2] 숫자, 연산자 분리 기능 추가

loadkrnis committed:
[step2] 연산 기능 & 다중 연산 기능 테스트 추가

loadkrnis committed:
[step2] 연산 기능 & 다중 연산 기능 추가

loadkrnis committed:
[step2] 에러 처리 테스트 추가

loadkrnis committed:
[step2] 에러 처리 기능 추가
 - 나누기 0 에러 처리 추가

loadkrnis commented on src/test/kotlin/step2/calculator/SplitterTest.kt:
```diff
@@ -0,0 +1,186 @@
+  }
+
+  describe("에러 처리") {
+//    it("입력 값이 null 인 경우 IllegalArgumentException 를 던진다.") {
+//      // given
+//      val input: String = null!!
+//
+//      // when
+//      val exception = shouldThrow<IllegalArgumentException> {
+//        Splitter(input)
+//      }
+//
+//      // then
+//      exception.message shouldBe "유효하지 않은 입력 값입니다."
+//    }
```
이 테스트가 의미있는 테스트인지 잘 모르겠어서 주석처리를 했습니다.
혹시 이 테스트에 대해서는 어떻게 생각하실까욤? 😅

hyunniiii replied:
저는 충분히 의미있는 테스트라고 생각합니다.
입력값이 null일 경우 에러를 던지는 상황에 대한 테스트는 필요한 것 같아요.

다만 
```kotlin
    it("입력 값이 null 인 경우 IllegalArgumentException 를 던진다.") {
      val input: String? = null

      val exception = shouldThrow<IllegalArgumentException> {
        Splitter(input)
      }
      exception.message shouldBe "유효하지 않은 입력 값입니다."
    }
```

이렇게 작성되어야 의미있는 테스트가 될 것 같습니다!

loadkrnis committed:
[step2] 린트 적용

hyunniiii commented on README.md:
```diff
@@ -1 +1,19 @@
\ No newline at end of file
+# kotlin-racingcar
+
+## [step 2] 문자열 사칙 연산 계산기
```
요구사항 작성 👍

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,15 @@
+        val result = operator.calculate(splitter.left, splitter.right)
+
+        if (splitter.hasOther) {
+            return execute("$result ${splitter.other}")
```
연산된 값을 이용해서 다시 `String` 형태로 조합하고 또 연산을 해주고 있는 것 같아요.
이렇게 쪼개고, 조합하기보다는 쪼개진 값을 컬렉션에 담고 컬렉션에 담긴 값을 이용해서 바로 연산하도록 해보는건 어떨까요?

hyunniiii commented on src/main/kotlin/step2/calculator/Math.kt:
```diff
@@ -0,0 +1,16 @@
+package step2.calculator
+
+class Math {
```
상태를 가지고 있지 않으니 `object`로 선언해보는건 어떨까요?
https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview

hyunniiii commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,13 @@
+package step2.calculator
+
+enum class Operator(private val operator: String, val calculate: (Long, Long) -> Long) {
```
enum 사용 👍

hyunniiii commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,13 @@
+    DIVIDE("/", { a, b -> Math.divide(a, b) });
+
+    companion object {
+        fun findByString(string: String): Operator =
```
파라미터 이름을 `string` 보다는 조금 더 의미있는 이름으로 변경해주는게 좋을 것 같아요 ″̮ 
참고로 변수명에 되도록이면 자료형은 적어주지 않는게 좋습니다. 

https://tecoble.techcourse.co.kr/post/2020-04-24-variable_naming/

hyunniiii commented on src/main/kotlin/step2/calculator/Splitter.kt:
```diff
@@ -0,0 +1,19 @@
+    val other: String
+
+    init {
+        expression.split(" ").let {
```
`" "` 와 같은 매직넘버는 의미를 알 수 있도록 상수화 해주시는게 좋습니다 ″̮ 
https://javabom.tistory.com/28

hyunniiii commented on src/test/kotlin/step2/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,64 @@
+import io.kotest.data.row
+import io.kotest.matchers.shouldBe
+
+class CalculatorTest : DescribeSpec({
```
테스트를 README.md에 적어주신 구현목록을 이용해서 작성해보는건 어떨까요?

`Calculator`의 테스트는 `이항식, 삼항식, 연산자가 같을 경우, 연산자가 다를 경우` 이러한 상황보다는 `연산 기능`이나 `다중 연산`에 대해 초점을 맞추고 작성되어야할 것 같아요.

예를 들어,
1. 올바른 문자열이 입력되었을 경우
2. 올바르지 않은 문자열이 입력되었을 경우

이러한 상황에서 어떠한 결과를 기대하는지에 대해 테스트를 작성해보는건 어떨까요?

hyunniiii commented on src/test/kotlin/step2/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,64 @@
+            row("8 * 4", 32),
+            row("8 / 4", 2)
+        ) { input, expectResult ->
+            // given
```
이러한 주석은 테스트 이름으로 표현해주시면 따로 작성안해주셔도 될 것 같습니다!

hyunniiii reviewed (changes requested):
안녕하세요 민규님~!
몇가지 코멘트 남겼으니 확인부탁드립니다 😃

hyunniiii commented:
질문주신 부분에 대해 답변 남깁니다. 

개발을 하다보면 코드 변경은 불가피하다고 생각합니다.
`public a()`가 다른 어느곳에서도 사용되지 않으면 `private a()`으로 변경해주겠죠?
`private a()`가 되는 순간 테스트하기가 어려워집니다.
하지만 `private a()`를 `public b()` 내부에서 호출하고 있다면 `public b()` 테스트에서 a 메서드의 로직을 검증할 수 있을 것 입니다.
이런식으로 상황에 따라 적절하게 변경해주시면 됩니다 ″̮ 

static, instance 메서드 부분은 https://mygumi.tistory.com/253 이 링크를 참고해보면 좋을 것 같습니다 😄 

또한 테스트에 대해 읽어보기 좋은글이 있어 링크 남겨드렸으니 한번 봐주시면 좋을 것 같습니다.
https://brunch.co.kr/@pubjinson/16

질문에 대한 답변이 되었는지 모르겠네요 😭 
이해가 안가시는 부분이 있다면 바로 질문 부탁드립니다 🙏

loadkrnis committed:
[step2] 1차 리뷰 반영
 - 입력 값이 null일 경우 에러 처리
 - 컬렉션에 담긴 값을 이용하여 연산하도록 수정 (메소드 오버로딩)
 - 상태가 없는 class -> object로 선언
 - 변수 네이밍 변경 string -> symbol, findByString -> find
 - given when then 삭제
 - 린트 적용

loadkrnis commented:
리뷰 잘 확인했습니다. :)
또한 답변 덕분에 상황에 따라 테스트 코드도 적절히 변경될 수 있다는 사실을 알게 되었습니다.
특히 첨부해주신 브런치 링크가 테스트 코드에 대해 많은 생각을 하게 된 것 같습니다!
감사합니다 😉

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,20 @@
+package step2.calculator
+
+class Calculator {
+
```
의미 없는 공백은 최대한 없애주시는게 좋습니다 ″̮

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,20 @@
+
+class Calculator {
+
+    fun execute(input: String): Long {
```
`execute`는 private 한 `execute`를 호출하는 용도인 것 같아요.
하나의 `execute`만 사용하도록 해보는건 어떨까요?
오버로딩된 메서드로 인해 한눈에 파악하기 어려운 것 같습니다.

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,20 @@
+class Calculator {
+
+    fun execute(input: String): Long {
+        return this.execute(input.split(" "))
```
`" "` 매직넘버가 있네요.
상수화로 의미를 나타내주는건 어떨까요?

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,20 @@
+
+    private fun execute(input: List<String>): Long {
+        val splitter = Splitter(input)
+        val operator = splitter.operator
```
`operator`는 예약어입니다.
예약어는 변수명으로 사용하는것을 지양해주세요 ″̮

hyunniiii commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,13 @@
+    DIVIDE("/", { a, b -> Math.divide(a, b) });
+
+    companion object {
+        fun find(symbol: String): Operator =
```
이 부분도 테스트할 수 있지 않을까요?

hyunniiii commented on src/main/kotlin/step2/calculator/Splitter.kt:
```diff
@@ -0,0 +1,20 @@
+    val other: List<String>
+
+    init {
+        expression ?: throw IllegalArgumentException("유효하지 않은 입력 값입니다.")
```
입력 받은 값이 `null`인 경우를 검증하는 책임이 `Splitter`의 책임인지 고민해보면 좋을 것 같아요 : )

loadkrnis replied:
이 부분은 .. 사실 고민 후 에러처리를 추가한 것입니다 ㅠㅠ... 🥲
분할할 때 잘못되었으면 에러를 던지는게 `Splitter` 의 책임이라고 생각했는데요.
혹시 리뷰어님의 의견은 어떠실까용? 🤔

hyunniiii replied:
`분할 시 잘못되었을 때 에러를 던진다.` 보다는 `이미 분할된 값이 null인지 아닌지`를 검증하는 부분인 것 같습니다.
분할할 때 에러를 던지는것은 `String` 값을 `List<String>`으로 분할할때 해주어야하지 않을까요? ″̮

hyunniiii commented on src/main/kotlin/step2/calculator/StringUtil.kt:
```diff
@@ -0,0 +1,4 @@
+package step2.calculator
+
+fun String.toLongOrThrow(): Long =
```
`Splitter`에서만 사용되니 `Splitter` 내부에 선언해줘도 괜찮을 것 같습니다. ″̮

hyunniiii replied:
사용하지 않는 코드는 삭제해주세요 ″̮

hyunniiii reviewed (changes requested):
안녕하세요 😄 
피드백 잘 반영해주셨네요.
몇가지 코멘트 남겼으니 확인부탁드립니다.
다음번 요청은 빠른 미션 진행을 위해 확인 후 머지하겠습니다!

loadkrnis committed:
[step2] 리뷰 반영

loadkrnis committed:
[step2] 린트 적용

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+package step2.calculator
+
+class Calculator {
+    companion object {
+        const val DELIMITER = " "
+    }
```
kotlin convention에 맞도록 리팩터링 해주면 어떨까요?

1.Property declarations and initializer blocks

2.Secondary constructors

3.Method declarations

4.Companion object

https://kotlinlang.org/docs/coding-conventions.html#class-layout

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+        const val DELIMITER = " "
+    }
+
+    fun execute(input: String): Long {
```
이 execute는 없애도 좋을 것 같아요 😄

loadkrnis replied:
저 메소드는 계산기 프로그램 시작점이라 남겨두었는데 없애야할까요? 🤔

hyunniiii commented on src/main/kotlin/step2/calculator/Calculator.kt:
```diff
@@ -0,0 +1,23 @@
+
+    private fun calculate(input: List<String>): Long {
+        val splitter = Splitter(input)
+        val operator = splitter.arithmeticOperator
```
`operator`는 kotlin 예약어입니다.
사용을 최대한 지양해주세요 ″̮

hyunniiii commented on src/main/kotlin/step2/calculator/Operator.kt:
```diff
@@ -0,0 +1,13 @@
+package step2.calculator
+
+enum class Operator(private val symbol: String, val calculate: (Long, Long) -> Long) {
+    PLUS("+", { a, b -> Math.plus(a, b) }),
```
```suggestion
    PLUS("+", { a, b -> a + b }),
```
이렇게하면 코드 양이 더 적을것 같습니다.
복잡도가 높지 않으니 추출하지 않아도 괜찮을 것 같아요.

hyunniiii reviewed (approved):
안녕하세요 민규님~!
몇가지 코멘트 남겼습니다.
다음 미션과 함께 반영 부탁드립니다 😄
